src/dbnode/storage/index.go

```diff
@@ -909,17 +909,25 @@ func (i *nsIndex) AggregateQuery(
 ) (index.AggregateQueryResult, error) {
 	// Get results and set the filters, namespace ID and size limit.
 	results := i.aggregateResultsPool.Get()
-	results.Reset(i.nsMetadata.ID(), index.AggregateResultsOptions{
-		SizeLimit:  opts.Limit,
-		TermFilter: opts.TermFilter,
-		Type:       opts.Type,
-	})
+	aopts := index.AggregateResultsOptions{
+		SizeLimit:   opts.Limit,
+		FieldFilter: opts.FieldFilter,
+		Type:        opts.Type,
+	}
 	ctx.RegisterFinalizer(results)
 	// use appropriate fn to query underlying blocks.
+	// default to block.Query()
 	fn := i.execBlockQueryFn
+	// use block.Aggregate() when possible
 	if query.Equal(allQuery) {
 		fn = i.execBlockAggregateQueryFn
 	}
+	field, isField := idx.FieldQuery(query.Query)
+	if isField {
+		fn = i.execBlockAggregateQueryFn
+		aopts.FieldFilter = append(index.AggregateFieldFilter{field}, aopts.FieldFilter...)
```

Comment:
Hm only question here, is for the default query coming from query engine - don't we set both the FieldFilter and the TagMatcher turns into a FieldQuery? 

If so, would we add the same field twice to the field filter?  Would that mean iterating that field twice instead of once?  After reading the field iterator code it wasn't clear, also I'd need to trace the query engine code again, but I though it both sent a tag matcher that is a FieldQuery and might also set the FieldFilter too.

Reply:
Hm, seems like it does (from `tag_values.go`):
```go
	return &storage.CompleteTagsQuery{
		// NB: necessarily spans the entire timerange for the index.
		Start:            time.Time{},
		End:              h.nowFn(),
		CompleteNameOnly: false,
		FilterNameTags:   [][]byte{nameBytes},
		TagMatchers: models.Matchers{
			models.Matcher{
				Type:  models.MatchRegexp,
				Name:  nameBytes,
				Value: matchValues,
			},
		},
	}, nil
```

Reply:
You're right - it would have been a duplicate iteration. Wouldn't have affected correctness but it's redundant. 

Made the change to only add the field filter if it doesn't exist in the list already, and further to de-dupe the filters to avoid this if users do something silly too.